### PR TITLE
feat(config): new config redirectAllowList

### PR DESCRIPTION
**Pull request by Matthew-Mallimo**
## Description

This adds a new `appConfig` option to configure an allow list for redirects.

## Motivation and Context

We dont want to redirect anyone to malicious URLs in the case of user-inputed redirects.

## How Has This Been Tested?

Unit and locally

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (adding or updating documentation)
- [ ] Dependency update
- [ ] Security update

## Checklist:

- [x] My change requires a change to the documentation and I have updated the documentation accordingly.
- [x] These changes should be applied to a maintenance branch.
- [ ] This change requires cross browser checks.
- [ ] Performance tests should be ran against the server prior to merging.
- [ ] This change impacts caching for client browsers.
- [ ] This change impacts HTTP headers.
- [ ] This change adds additional environment variable requirements for One App users.
- [ ] I have added the Apache 2.0 license header to any new files created.

## What is the Impact to Developers Using One App?

Allows further security enhancements through a customizable allow list.

### CODE REVIEW

File: __tests__/server/middleware/checkStateForRedirect.spec.js [hunk 1]
```diff
@@ -40,4 +49,14 @@ describe('checkStateForRedirect', () => {
     expect(res.redirect).not.toHaveBeenCalled();
     expect(next).toHaveBeenCalled();
   });
+  it('should call next with an error if the redirect URL is not in the allow list', () => {
+    const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => null);
+    validateRedirectUrl.mockImplementationOnce(() => false);
+    state = fromJS({ redirection: { destination } });
+    checkStateForRedirect(req, res, next);
+    expect(res.redirect).not.toHaveBeenCalled();
+    expect(next).toHaveBeenCalled();
```
Comment (bishnubista) [inline __tests__/server/middleware/checkStateForRedirect.spec.js:58]: Shouldn't we clear mocks or add `toHaveBeenCalledTimes`?

**Review by bishnubista — DISMISSED** [review]

**Review by dogpatch626 — DISMISSED** [review]

**Review by 10xLaCroixDrinker — CHANGES_REQUESTED** [review]
It looks right now like the only way to revert the enabling this feature is to deploy a new module release that sets the value to `[]`. Please refactor so that the change can be rolled back to where this feature is not configured and will be disabled.

File: src/server/middleware/createRequestHtmlFragment.jsx [hunk 2]
```diff
@@ -117,6 +121,11 @@ export default function createRequestHtmlFragment({ createRoutes }) {
         });
 
         if (redirect) {
+          if (!validateRedirectUrl(redirect.url)) {
```
[the remark this answers is not in the document]
Comment (Matthew-Mallimo) [reply src/server/middleware/createRequestHtmlFragment.jsx]: Changed in 1f920a4e30ee9fa88371d9a2516f1348fa31d30b

**Comment by Matthew-Mallimo** [comment]
> It looks right now like the only way to revert the enabling this feature is to deploy a new module release that sets the value to `[]`. Please refactor so that the change can be rolled back to where this feature is not configured and will be disabled.

Addressed in 1f920a4e30ee9fa88371d9a2516f1348fa31d30b

**Review by bishnubista — APPROVED** [review]

**Review by 10xLaCroixDrinker — APPROVED** [review]

File: src/server/utils/onModuleLoad.js [hunk 3]
```diff
@@ -91,6 +87,45 @@ export default function onModuleLoad({
...
+  validateCspIsPresent(csp);
+  clearModulesUsingExternals();
+  if (provideStateConfig) {
+    setStateConfig(provideStateConfig);
+  }
+  if (errorPageUrl) {
+    setErrorPage(errorPageUrl);
+  }
+  if (redirectAllowList) {
+    setRedirectAllowList(redirectAllowList);
```
Comment (JAdshead) [inline src/server/utils/onModuleLoad.js:103]: what happens when a root module has a redirect allow list, then a new version gets released without a redirect allow list? this needs to ensure that its reset

File: src/server/utils/redirectAllowList.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+// Spread it to keep it immutable
+export const getRedirectAllowList = () => [...redirectAllowList];
+
+export const isRedirectUrlAllowed = (url) => {
+  // If a redirect list is not configured, allow the redirect.
+  // This prevents the redirect allow list from becoming a breaking change.
+  if (redirectAllowList.length === 0 || isAllowedUrl(url)) {
+    return true;
+  }
+  return false;
```
Comment (JAdshead) [inline src/server/utils/redirectAllowList.js:37]:
```suggestion
  // If a redirect list is not configured, allow the redirect.
  // This prevents the redirect allow list from becoming a breaking change.
  return redirectAllowList.length === 0 || isAllowedUrl(url)
```

File: __tests__/server/utils/onModuleLoad.spec.jsx [hunk 5]
```diff
@@ -180,7 +184,31 @@ describe('onModuleLoad', () => {
     });
     expect(setStateConfig).toHaveBeenCalledWith(provideStateConfig);
   });
-
+  it('calls setStateConfig if setRedirectAllowList is supplied', () => {
```
Comment (JAdshead) [inline __tests__/server/utils/onModuleLoad.spec.jsx:187]:
```suggestion
  it('calls setRedirectAllowList if setRedirectAllowList is supplied', () => {
```

File: __tests__/server/utils/onModuleLoad.spec.jsx [hunk 6]
```diff
@@ -180,7 +184,31 @@ describe('onModuleLoad', () => {
...
+          redirectAllowList: ['https://americanexpress.com'],
+          csp,
+        },
+        [META_DATA_KEY]: { version: '1.0.5' },
+      },
+      moduleName: 'some-root',
+    });
+    expect(setRedirectAllowList).toHaveBeenCalledWith(['https://americanexpress.com']);
+  });
+  it('calls setStateConfig with empty array if setRedirectAllowList is NOT supplied', () => {
```
Comment (JAdshead) [inline __tests__/server/utils/onModuleLoad.spec.jsx:200]:
```suggestion
  it('calls setRedirectAllowList with empty array if setRedirectAllowList is NOT supplied', () => {
```

File: src/server/utils/onModuleLoad.js [hunk 7]
```diff
@@ -91,6 +87,45 @@ export default function onModuleLoad({
       pwa,
       errorPageUrl,
       dnsCache,
+      redirectAllowList,
```
Comment (JAdshead) [inline src/server/utils/onModuleLoad.js:90]:
```suggestion
      redirectAllowList = [],
```

**Review by code-forger — APPROVED** [review]

**Review by JAdshead — APPROVED** [review]

**Pull request merged**